File: csharp/src/IceBox/Service.cs
// Copyright (c) ZeroC, Inc.

#nullable enable

namespace IceBox;

/// <summary>
/// The exception that is thrown when an IceBox service fails to start.
/// </summary>
/// <remarks>You can throw any exception from your implementation of <see cref="Service.start"/>. This exception is
/// provided for backward compatibility with earlier versions of IceBox.</remarks>
public sealed class FailureException : Ice.LocalException
{
    public string reason => Message;

    public FailureException(string? message, System.Exception? innerException = null)
        : base(message, innerException)
    {
    }

    public override string ice_id() => "::IceBox::FailureException";
}

/// <summary>
/// Represents an IceBox service that you implement and that the IceBox service manager starts and later stops.
/// The same service can be started and stopped multiple times.
/// </summary>
public interface Service
{
    /// <summary>
    /// Starts the service.
    /// </summary>
    /// <param name="name">The service's name, as specified in configuration.</param>
    /// <param name="communicator">A communicator for use by the service. The IceBox service manager creates this
    /// communicator when it starts, and destroys this communicator when it shuts down.</param>
    /// <param name="args">The service arguments that were not converted into properties of <paramref name="communicator"/>.</param>
    void start(string name, Ice.Communicator communicator, string[] args);

    /// <summary>
    /// Stops the service.
    /// </summary>
    void stop();
}

File: java/src/com.zeroc.icebox/src/main/java/com/zeroc/IceBox/Service.java
// Copyright (c) ZeroC, Inc.

package com.zeroc.IceBox;

import com.zeroc.Ice.Communicator;

/** An application service managed by a {@link ServiceManager}. */
public interface Service {
    /**
     * Start the service. The given communicator is created by the {@link ServiceManager} for use by
     * the service. This communicator may also be used by other services, depending on the service configuration.
     *
     * <p class="Note">The {@link ServiceManager} owns this communicator, and is responsible for destroying it.
     *
     * @param name the service's name, as determined by the configuration
     * @param communicator a communicator for use by the service
     * @param args the service arguments that were not converted into properties
     * @throws FailureException if {@link #start} failed.
     */
    void start(String name, Communicator communicator, String[] args);

    /** Stop the service. */
    void stop();
}
